[BryanL2303] iP

BryanL2303 opened the pull request:
# Klee

> Hello, are you here for playtime with Klee?

Klee is an adorable and energetic child from the game Genshin Impact.
She will help you write things down in her book so that you won't forget them.

- text based
- easy to learn
- Klee is ~~fast~~ SUPER FAST 

All you need to do is,

1. download the file Klee.jar [here](https://github.com/BryanL2303/ip/releases/tag/A-Jar).
2. open command prompt and go to the directory which contains the file above.
3. Run the program by running > java -jar Klee.jar
4. Let Klee manage your tasks for you 😄 

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Throw bombs (Maybe? soon? _Probably not_)

_______________________________________________________________________________________________________

If you are a Java programmer, you can check the repository out. Here's the ```main``` function

```java
public class Klee {
    public static void main(String[] args) {
        Ui ui = new Ui();
        Parser parser = new Parser();
        Storage storage = new Storage();

        TaskList tasks = storage.retrieveTasks(ui);

        ui.showWelcome();

        run(ui, parser, storage, tasks);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

BryanL2303 committed:
Completed level 0

BryanL2303 committed:
Completed level 1

BryanL2303 committed:
Completed level 2

BryanL2303 committed:
Added basic functions for Task class

BryanL2303 committed:
Completed level 3

BryanL2303 committed:
Completed level 4

BryanL2303 committed:
Added basic text testing

BryanL2303 committed:
Completed level 5 with custom exception

BryanL2303 committed:
Completed level 6

BryanL2303 committed:
Replace nested if/else with enums

The if/else nest was too deep.

Using enums for instructions is easier for bug catching and future
expansions.

The instructions enum is declared at the top of the file
The nested if/else is written as a switch case instead.

BryanL2303 committed:
Add type for Task class

Each task will type their own type into the status.

To save the tasks into the data file we need to be able to access the
type indicator for each task.

Declare type into Task class and instantiate them within each inherited
class.

Now each task can access their own types.

BryanL2303 committed:
Add function to retrieve type of Task

Previous commit required this function.

The function has now been added so each Task can access their own type
for writting data.

BryanL2303 committed:
Add function to convert task to textline

To convert the task into 1 line of txt to save the tasks easily.

Deadline and Event class overwrites the function to add the additional
parameters.

BryanL2303 committed:
Add function to convert textline to Task

To simplify the process for creating Tasks after reading from the txt
file.

BryanL2303 committed:
Add feature to save tasks

Whenever the program stops running the tasks will be lost.

To ensure that the list of tasks is kept, save the tasks into a local
txt file and read it the next time the program is run.

BryanL2303 committed:
Merge branch 'branch-Level-7'

BryanL2303 committed:
Change functions to recognise date and time

Variables such as the deadline or start and end of events are stored as
Strings.

It makes more sense to store them as LocalDateTime for logical
validation.

Changed the declaration of the classes involved from String to
LocalDateTime.
Added functions in main class to convert between String and
LocalDateTime and replaced parts of the function to save and load from
the text file.

BryanL2303 committed:
Merge branch 'branch-Level-8'

BryanL2303 committed:
Extract functions into Packages and Objects

Almost all functions were defined in the main Klee class.

This is very untidy and makes it difficult to navigate the code.

Extract functions into Objects and put similar classes into packages.
command package includes all commands that Klee accepts.
tasks package includes all types of tasks that Klee can note down.
General commands are split into general classes within Klee package.

BryanL2303 committed:
Merge branch 'branch-A-MoreOOP'

BryanL2303 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

BryanL2303 committed:
Update path to main class in gradle build file

Updated from the template give to fit the project.

BryanL2303 committed:
Add equals function to classes

To facilitate testing functions, each class needs a way to identify if
instances are equal.

equals function has been overridden in all classes under command and
KleeException.

Parser has been debugged for typos in creating KleeException.

BryanL2303 committed:
Add unit tests for TaskList

Override equals for all task classes to simplify TaskList testing.

Create TaskListTest class with tests for all functions

BryanL2303 committed:
Merge branch 'branch-JUnit'

BryanL2303 committed:
Update gradle to run shadowJar

Compile the class Klee into .jar file so that we can run the program
from the command prompt.

BryanL2303 committed:
Add JavaDoc for all classes and methods.

Add Javadoc for all classes and methods for easier understanding.
Removed functions with no usage.

BryanL2303 committed:
Fix code to align with coding standard

Readjusted code to align with coding standard.
No functions were changed.

BryanL2303 committed:
Add new command Find

Add function to filter task descriptions with a search term in new
Command Find.
Updated Ui and Parser to recognise this new command and output the
appropriate response.

BryanL2303 committed:
Add JavaDoc to new methods

Add JavaDoc that was left out.

BryanL2303 committed:
Merge branch 'branch-A-JavaDoc'

BryanL2303 committed:
Merge branch 'branch-A-CodingStandard'

BryanL2303 committed:
Merge branch 'branch-Level-9'

BryanL2303 committed:
Fix merge conflict

Previously a new function createed in a branch was wrongly removed when
handling merge conflict.

The missing function is now added back.

BryanL2303 committed:
Add files to set up style checker

To run script to check if code follows course coding standards.

Created files with checkstyles and updated build.gradle file.

BryanL2303 committed:
Fix coding style for Ui class

Fix errors found by style checker.

BryanL2303 committed:
Correct errors with coding standard

Ran the checkstyle tests and fixed all issues spotted.

BryanL2303 committed:
Merge branch 'branch-A-CheckStyle'

BryanL2303 committed:
Fix typo in function

The Unmark command is wrongly calling showMarked instead of
showUnMarked.

Corrected the wrong function call.

BryanL2303 committed:
Recode main class and Ui to create GUi

Instead of having a text based Ui, generate a GUi using JavaFX.

Recode the entire Ui class and added function to echo user input.
All functions are now changed to output to GUi rather than command line.

Main class was recoded to set up stage and handle commands onAction.
Added Launcher class to launch GUi, app should now be run from Launch
class rather than Klee class.

BryanL2303 committed:
Edit build.gradle to change seed and base file

Compile Launcher.jar file.

BryanL2303 committed:
Merge branch 'branch-Level-10'

BryanL2303 committed:
Update README.md

The README was not updated.

Update with current status.

tzyian commented on src/main/java/klee/Parser.java:
```diff
@@ -0,0 +1,183 @@
+    public Parser() {}
+
+    /**
+     * Parse the String dateTime into a LocalDateTime instance.
+     * If the syntax of dateTime cannot be understood throw KleeException.
```
Following the code standard, should this be plural?
Likewise for other parts

tzyian commented on src/main/java/klee/Parser.java:
```diff
@@ -0,0 +1,183 @@
+     * @throws KleeException
+     */
+    public static LocalDateTime parseDateTime(String dateTime) throws KleeException {
+        String[] splitDateTime = dateTime.split(" ");
```
Should this be plural?
Likewise for other parts

tzyian commented on src/main/java/klee/task/Task.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @return If the task is done.
+     */
+    public boolean done() {
```
Should method name be of this format?

tzyian commented on src/main/java/klee/Parser.java:
```diff
@@ -0,0 +1,183 @@
+                            return new Mark(index);
+                        } else {
+                            throw new KleeException("Is that a negative number? "
+                                    + "I usually start counting from number 1...");
```
Would it be clearer to extract the cases and nested if-elses into separate methods?

tzyian commented on src/main/java/klee/Parser.java:
```diff
@@ -0,0 +1,183 @@
+                    throw new KleeException("There should be an integer after mark to indicate which task to mark!");
+                }
+            case "unmark":
+                if (command.length == 2) {
```
Should you add newlines between cases to demarcate changes in logic?

tzyian reviewed (commented):
A few coding standard violations, but LGTM!

BryanL2303 committed:
Fix coding standard violation

As pointed by PR reviewer.

All issues has been addressed to fit with coding standards.

BryanL2303 committed:
Merge branch 'master' of https://github.com/BryanL2303/ip

BryanL2303 committed:
Add assert statements

For relevant method calls, assert the returned value.

BryanL2303 committed:
Refactor code to be more readable

Case statement has too many nested functions for each case.

Each case has been refactored into their own function and the case
statement calls to each of those functions.

Return statements for boolean with too many conditions are separated and
named appropriately.

BryanL2303 committed:
Merge branch 'branch-A-Assertions'

BryanL2303 committed:
Merge branch 'branch-A-CodeQuality'

BryanL2303 committed:
Add workflow file for github

Added .yml file as provided by the course, removed IO testing on shell.

BryanL2303 committed:
Merge branch 'branch-A-CI'

laybach commented on src/main/java/klee/Parser.java:
```diff
@@ -0,0 +1,260 @@
+     * @return The Command Unmark if the input is valid
+     * @throws KleeException
+     */
+    public static Command caseUnmark(String[] inputComponents) throws KleeException {
```
good use of plural for an array

laybach commented on src/main/java/klee/command/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+     * @return If obj has the same fields.
+     */
+    @Override
+    public boolean equals(Object obj) {
```
should be isEqual() instead of equals() for boolean naming conventions

laybach commented on src/main/java/klee/command/Event.java:
```diff
@@ -0,0 +1,65 @@
+    @Override
+    public boolean equals(Object obj) {
+        if (obj.getClass() == Event.class) {
+            boolean hasSameFields = this.description.equals(((Event) obj).description) && this.from.equals((((Event) obj).from))
```
nicely done for boolean variable naming

laybach commented on src/main/java/klee/task/Event.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    public String toText() {
+        return super.toText() + " / " + from.format(DateTimeFormatter.ofPattern("yyyy MM dd H m")) + " / "
+                + to.format(DateTimeFormatter.ofPattern("yyyy MM dd H m"));
```
Great to see you continue your code in the next line to prevent going over 120 characters per line

laybach reviewed (commented):
LGTM! Code looks well written and mostly follows the coding standards.

BryanL2303 committed:
Merge branch 'master' into branch-Level-10

BryanL2303 committed:
Add profile images

Rather than only having text in the dialogs which looks very plain,
added images in a resourse folder which will be displayed to visualise
the conversation between user and the program.

Updated the outputs on the Ui class.
Removed 1 assert line on the Storage class causing bugs.
Editted the main class to increase the size of the GUi.
Added images in a resourses/images folder.

BryanL2303 committed:
Merge branch 'branch-Level-10'

BryanL2303 committed:
Fix codestyle error

Fixed errors pointed out by running build.

BryanL2303 committed:
Add user guide and screenshots

Updated user guide

Included representative screenshot as well as other screenshots for the
user guide.

Updated Ui to display correct image for list command

BryanL2303 committed:
Update gradle file for shadowJar

Rename the jar file compile to Klee instead of Launcher

BryanL2303 committed:
Improve code quality

Rename some javadoc headers to fit format.

Abstracted one function from parseDateTime to shorten the method.

BryanL2303 committed:
Fix code style violation

BryanL2303 committed:
Update README.md

Provide information post project completion
